[8.0][event_registration_partner_unique] Unique partners in registrations

yajo opened the pull request:
This module allows you to avoid having a partner registrated twice in the same event.

@rafaelbn

yajo committed:
[8.0][event_registration_partner_unique] Unique partners in registrations.

This module allows you to avoid having a partner registrated twice in the same event.

rafaelbn commented:
@Yajo by default Odoo doesn't have Partners in Events Registrations

![29-02-2016 18-12-58](https://cloud.githubusercontent.com/assets/8736623/13402342/1abd0ec0-df10-11e5-85c3-b6ea65509f3f.png)

In fact it's done in OCA by https://github.com/OCA/event/tree/8.0/partner_event (we show add this restriction in this module and maybe could be an improvement)

In this PR how do you add Partners in Event registration without dependencies?

yajo commented:
You have to press the top-right button "Registrations"

rafaelbn commented on event_registration_partner_unique/__openerp__.py:
```diff
@@ -0,0 +1,18 @@
+    "installable": True,
+    "pre_init_hook": "pre_init_hook",
+    "depends": [
+        "event",
```
It doesn't depends in https://github.com/OCA/event/tree/8.0/partner_event?

yajo replied:
That module just adds a shortcut. See https://github.com/OCA/event/pull/38#issuecomment-190296383

rafaelbn commented:
Perfect, tested :+1:

carlosdauden commented:
:+1:

andhit-r commented:
Hi @Yajo ,

I try:
1. I set registration max to 1
   ![registration_unique_3](https://cloud.githubusercontent.com/assets/1947376/13481456/69b031ec-e118-11e5-96ad-75dfddec9a12.png)
2. Create sale order with purchase of 2 ticket
   ![registration_unique_2](https://cloud.githubusercontent.com/assets/1947376/13481450/61958da4-e118-11e5-845c-c0f5372c260f.png)
3. When i confirm sale warning shows
   ![registration_unique](https://cloud.githubusercontent.com/assets/1947376/13481467/7b87a6c0-e118-11e5-94d8-b4b4583d01a8.png)

Warning message still not correct (it shows %d instead od max purchase information)

andhit-r commented:
I try:
1. Set event max participant per registration to 1
   ![registration_unique_3](https://cloud.githubusercontent.com/assets/1947376/13481537/13c34e58-e119-11e5-939d-cef25c172e88.png)
2. I create and confirm SO for 1 participant
   ![registration_unique_4](https://cloud.githubusercontent.com/assets/1947376/13481550/2638dfbc-e119-11e5-97b6-0dcfd4c50b09.png)
3. I create SO for same partner
   ![registration_unique_5](https://cloud.githubusercontent.com/assets/1947376/13481558/3c485d64-e119-11e5-824c-a75011dbbad1.png)

When i confirm second quotation i got this error

Traceback (most recent call last):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/http.py", line 537, in _handle_exception
    return super(JsonRequest, self)._handle_exception(exception)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/http.py", line 574, in dispatch
    result = self._call_function(*_self.params)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/http.py", line 310, in _call_function
    return checked_call(self.db, *args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/service/model.py", line 118, in wrapper
    return f(dbname, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/http.py", line 307, in checked_call
    return self.endpoint(_a, *_kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/http.py", line 803, in __call__
    return self.method(_args, *_kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/http.py", line 403, in response_wrap
    response = f(_args, *_kw)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/addons/web/controllers/main.py", line 948, in call_button
    action = self._call_kw(model, method, args, {})
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/addons/web/controllers/main.py", line 936, in _call_kw
    return getattr(request.registry.get(model), method)(request.cr, request.uid, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 268, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/addons/sale/sale.py", line 609, in action_button_confirm
    self.signal_workflow(cr, uid, ids, 'order_confirm')
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 268, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/models.py", line 3571, in signal_workflow
    result[res_id] = workflow.trg_validate(uid, self._name, res_id, signal, cr)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/__init__.py", line 85, in trg_validate
    return WorkflowService.new(cr, uid, res_type, res_id).validate(signal)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/service.py", line 91, in validate
    res2 = wi.validate(signal)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/instance.py", line 80, in validate
    wi.process(signal=signal, force_running=force_running, stack=stack)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/workitem.py", line 120, in process
    ok = self._split_test(activity['split_mode'], signal, stack)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/workitem.py", line 248, in _split_test
    self._join_test(t[0], t[1], stack)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/workitem.py", line 257, in _join_test
    WorkflowItem.create(self.session, self.record, activity, inst_id, stack=stack)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/workitem.py", line 95, in create
    workflow_item.process(stack=stack)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/workitem.py", line 116, in process
    if not self._execute(activity, stack):
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/workitem.py", line 162, in _execute
    returned_action = self.wkf_expr_execute(activity)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/workitem.py", line 313, in wkf_expr_execute
    return self.wkf_expr_eval_expr(activity['action'])
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/workflow/workitem.py", line 291, in wkf_expr_eval_expr
    result = eval(line, env, nocopy=True)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/tools/safe_eval.py", line 314, in safe_eval
    return eval(c, globals_dict, locals_dict)
  File "", line 1, in <module>
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 266, in wrapper
    return new_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 592, in new_api
    result = method(self._model, cr, uid, self.ids, _args, *_old_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/addons/sale/sale.py", line 624, in action_wait
    self.pool.get('sale.order.line').button_confirm(cr, uid, [x.id for x in o.order_line if x.state != 'cancel'])
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 268, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/addons/event_sale/event_sale.py", line 128, in button_confirm
    registration_id = registration_obj.create(cr, uid, dic, context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 268, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 372, in old_api
    result = method(recs, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/addons/partner_event/models/event_registration.py", line 40, in create
    return super(EventRegistration, self).create(vals)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 266, in wrapper
    return new_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 508, in new_api
    result = method(self._model, cr, uid, _args, *_old_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/addons/mail/mail_thread.py", line 381, in create
    thread_id = super(mail_thread, self).create(cr, uid, values, context=context)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 268, in wrapper
    return old_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 372, in old_api
    result = method(recs, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/models.py", line 4093, in create
    record = self.browse(self._create(old_vals))
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 266, in wrapper
    return new_api(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/api.py", line 508, in new_api
    result = method(self._model, cr, uid, _args, *_old_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/models.py", line 4231, in _create
    tuple([u[2] for u in updates if len(u) > 2])
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/sql_db.py", line 158, in wrapper
    return f(self, _args, *_kwargs)
  File "/srv/openerp/instances/openerp-oca-runbot/parts/odoo-extra/runbot/static/build/3142688-38-b984a4/openerp/sql_db.py", line 234, in execute
    res = self._obj.execute(query, params)
ValueError: "duplicate key value violates unique constraint "event_registration_unique_partner_event"
DETAIL:  Key (partner_id, event_id)=(42, 4) already exists.
" while evaluating
u'action_wait()'

yajo committed:
Fix error message.

yajo committed:
Depends now on event_sale to fix [this bug][1].

It has to declare unique indexes instead of constraints to work.

[1]: https://github.com/OCA/event/pull/38#issuecomment-191525863

yajo commented:
Hi @andhit-r.

I fixed https://github.com/OCA/event/pull/38#issuecomment-191524711, although that bug comes from the `event_registration_seat_limit` addon, but it's a very simple fix (8eb4fca).

I also fixed https://github.com/OCA/event/pull/38#issuecomment-191525863, but that required me to depend on `event_sale`, and I know somebody who will not like that...

Anyway it should work now.

rafaelbn commented:
Hi @Yajo but this is not correct
```
I also fixed #38 (comment), but that required me to depend on event_sale, and I know somebody who will not like that...
```
This module mustn't require event_sale, you can use Odoo to manager you events and registration but you don't want to see them.

I thinks it must be another module (bridge) with auto-install in the case you have installed event_registration_partner_unique and you install event_sale

rafaelbn commented:
Hi @Yajo but this is not correct

> I also fixed #38 (comment), but that required me to depend on event_sale, and I know somebody who will not like that...

This module mustn't require event_sale, you can use Odoo to manage your events and registrations but you don't want to sell them.

I thinks it must be another module (bridge) with auto-install in the case you have installed event_registration_partner_unique and you install event_sale

cc @pedrobaeza @antespi

antespi commented:
Thanks @Yajo for yourr PR, but IMHO I would prefer another technical approach:
1. Do not use SQL constrains but inheritable Python constrains `_check_partner_unique`
2. Define a new field `forbid_duplicates` in `event.event` allowing user to configure each event, which events allow duplicates and which not. Doing this you do not need `pre_init_hook` validation.
3. Create another addon (auto installable) that depends on `event_sale` and `event_registration_partner_unique` and inherits `_check_partner_unique` method taking care about tickets
4. Python over `forbid_duplicates` that checks that if `True` there is no already duplicates in this event.

What do you think?

yajo commented:
I went this way to keep it simpler, but indeed yours is better. Although it requires a complete refactoring and splitting in 2 modules.

yajo committed:
Refactor module.

- Remove dependency on event_sale.
- Add custom field that allows to choose in which events to disable duplicates.
- Add tests.
- Add custom exception to be used in upcoming submodule (for event_sale).

yajo commented:
Module is updated. I'll post the 2nd in a separate PR, it's being prepared.

andhit-r commented:
Tested on runbot :+1:

andhit-r commented:
@OCA/crm-sales-marketing-maintainers this PR is ready to be merged (3 reviews)
